python/phase-2.md

```diff
@@ -0,0 +1,830 @@
+    ) -> None:
+        pass
+
+    def _push(self) -> None:
+        pass
+
+    def _pop(self) -> None:
+        pass
+
+    await def run(self, fn: Callable[..., Union[Awaitable[T], T]]) -> T:
+        pass
+
+    def __enter__(self) -> CancellationScope:
+        pass
+
+    def __exit__(self) -> None:
+        pass
+
+    def cancel(self) -> None:
+        pass
+
+    @property
+    def cancelled(self) -> bool:
+        pass
+```
+
+Notes:
+
+* Unlike TypeScript, Python still allows you to explicitly cancel individual activities and child workflows if you
+  started them with `start_`
+  * The consistency of having everything that has a `start_` be cancellable has value. I still expect most will use the
+    `execute_` form though
+* The workflow will have a root scope
+* This will work using `with CancellationScope()` and similar and is powered by `contextvars`
+  * `run` is async because the callable may have an async result, but `with` is the preferred approach for explicit use
+* `start_activity`/`execute_activity`/`start_child_workflow`/`execute_child_workflow` will implicitly use the current
+  cancellation scope
+* Do we need extra capabilities such as:
+  * Ability to explicitly add cancellables?
+  * Ability to unset the current instead of just forcing a detached one each time you want to opt out?
+* Should we disallow the `parent` to even be set in the constructor?
+  * Thinking we probably shouldn't allow you to set that
+* Cancellation of a workflow or a scope causes all things started with that to be issued a cancel
+  * Originally we weren't gonna use cancellation scopes but instead reuse the `asyncio` cancel feature but that throws
+    inside of awaited things and we want to cancel whether awaited or not
+  * This means `asyncio` cancellation (and `shield` etc) affects only in-workflow tasks awaited on, _not_ the actual
+    underlying activity/child being executed
+    * We could disable these features of `asyncio` so people aren't confused by their concept of cancellation and ours,
+      but there is still value in cancelling just the tasks
+
+## Runtime
+
+### Workflow Stepped Execution
+
+Notes:
+
+* We are gonna use `asyncio`, but whether we have to use our own event loop remains to be seen
+  * Have shown at https://github.com/cretz/python-determinism-runner basically using the approach from
+    https://stackoverflow.com/a/29797709 that we can run a single step and enforce that we have yielded only at
+    workflow-acceptable spots
+* Do want to prevent use of `asyncio.sleep`, `asyncio.wait_for`, and similar that are based on system time?
+  * This is much clearer than replacing their implementations so that people aren't confused about why a 50ms wait does
+    not work
+  * We may not be able to block everything until we use a sandbox (see next section)
+  * Making alternatives is a tad ugly, e.g. one can use `asyncio.gather()` without timeout but they can't with timeout,
```

Comment:
Is this still true if we implement our own event loop?

Reply:
My fear is what may seem like a local operation can increase history with timer creation. But I am on the fence.